test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -266,26 +279,117 @@ func responsesConfigMap(name string, namespace *corev1.Namespace) *corev1.Config
+		return cluster.Status.Phase == string(clusterv1.ClusterPhaseProvisioned)
+	}, intervals)
+}
+
+// beforeClusterUpgradeTestHandler provides an unblocked function which returns true if the Cluster has controlplanev1.RollingUpdateInProgressReason in its
+// ReadyCondition.
+func beforeClusterUpgradeTestHandler(ctx context.Context, c client.Client, namespace, clusterName, version string, intervals []interface{}) {
+	runtimeHookTestHandler(ctx, c, namespace, clusterName, "BeforeClusterUpgrade", func() bool {
+		cluster := &clusterv1.Cluster{}
+		var unblocked bool
+
+		// First ensure the Cluster topology has been updated to the target Kubernetes Version.
+		Eventually(func() bool {
+			Expect(c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: clusterName}, cluster)).To(Succeed())
+			return cluster.Spec.Topology.Version == version
+		}).Should(BeTrue(), "BeforeClusterUpgrade blocking condition false: Cluster topology has not been updated to the target Kubernetes Version")
+
+		// Check if the Cluster is showing the RollingUpdateInProgress condition reason. If it has the update process is unblocked.
+		if conditions.IsFalse(cluster, clusterv1.ReadyCondition) &&
+			conditions.GetReason(cluster, clusterv1.ReadyCondition) == controlplanev1.RollingUpdateInProgressReason {
+			unblocked = true
+		}
+		return unblocked
+	}, intervals)
+}
+
+// afterControlPlaneUpgradeTestHandler provides an unblocked function which returns true if any MachineDeployment in the Cluster
+// has upgraded to the target Kubernetes version.
+func afterControlPlaneUpgradeTestHandler(ctx context.Context, c client.Client, namespace, clusterName, version string, intervals []interface{}) {
+	runtimeHookTestHandler(ctx, c, namespace, clusterName, "AfterControlPlaneUpgrade", func() bool {
+		var unblocked bool
+		mds := &clusterv1.MachineDeploymentList{}
+		Expect(c.List(ctx, mds, client.MatchingLabels{
+			clusterv1.ClusterLabelName:          clusterName,
+			clusterv1.ClusterTopologyOwnedLabel: "",
+		})).To(Succeed())
+
+		// If any of the MachineDeployments have the target Kubernetes Version, the hook is unblocked.
+		for _, md := range mds.Items {
+			if *md.Spec.Template.Spec.Version == version {
+				unblocked = true
+			}
+		}
+		return unblocked
+	}, intervals)
+}
+
+func runtimeHookTestHandler(ctx context.Context, c client.Client, namespace, clusterName, hookName string, condition func() bool, intervals []interface{}) {
+	// First check that the LifecycleHook has been called at least once.
+	Eventually(func() bool {
+		err := checkLifecycleHooks(ctx, c, namespace, clusterName, map[string]string{hookName: ""})
+		return err != nil
+	}, 120*time.Second).Should(BeTrue(), "%s has not been called", hookName)
+
+	// condition should consistently be false as the Runtime hook is returning "Failure".
+	Consistently(func() bool {
+		return condition()
+	}, 10*time.Second).Should(BeFalse(), fmt.Sprintf("%s hook blocking condition succeeded before unblocking", hookName))
```

Comment:
I think we have this inverted a bit atm. 

> hook blocking condition succeeded before unblocking

It's kind of correct but our condition func doesn't return if it's blocked it returns if it's unblocked.

Can we flip the return values of the funcs we pass in and then use the following?

```suggestion
	// condition should consistently be false as the Runtime hook is returning "Failure".
	Consistently(func() bool {
		return blockingCondition()
	}, 10*time.Second).Should(BeTrue(), fmt.Sprintf("%s hook blocking condition succeeded before unblocking", hookName))
```

Just feels a bit easier to think through (blocking condition seems a more straightforward concept as an unblocking condition). WDYT?